vertx-auth-mtls/src/main/java/io/vertx/ext/auth/mtls/impl/MutualTlsAuthenticationProvider.java

```
@@ -0,0 +1,93 @@
+import io.vertx.ext.auth.authentication.Credentials;
+import io.vertx.ext.auth.mtls.CertificateIdentityExtractor;
+import io.vertx.ext.auth.mtls.CertificateIdentityValidator;
+import io.vertx.ext.auth.mtls.CertificateValidator;
+
+import java.security.cert.Certificate;
+
+/**
+ * {@link AuthenticationProvider} implementation for mTLS (MutualTLS) authentication. With mTLS authentication
+ * both server and client exchange certificates and validates each other's certificates.
+ */
+public class MutualTlsAuthenticationProvider implements AuthenticationProvider
+{
+    private final CertificateValidator certificateValidator;
+    private final CertificateIdentityExtractor identityExtractor;
+    private final CertificateIdentityValidator identityValidator;
+
+    public MutualTlsAuthenticationProvider(CertificateValidator certificateValidator,
+                                           CertificateIdentityExtractor identityExtractor,
+                                           CertificateIdentityValidator identityValidator)
+    {
+        this.certificateValidator = certificateValidator;
+        this.identityExtractor = identityExtractor;
+        this.identityValidator = identityValidator;
+    }
+
+    @Override
+    public Future<User> authenticate(Credentials credentials)
+    {
+        if (!(credentials instanceof CertificateCredentials))
+        {
+            return Future.failedFuture("CertificateCredentials expected for mTLS authentication");
+        }
+
+        CertificateCredentials certificateCredentials = (CertificateCredentials) credentials;
+        String identity;
+        try
+        {
+            certificateValidator.isValidCertificate(certificateCredentials);
+            identity = identityExtractor.identity(certificateCredentials.certificateChain().toArray(new Certificate[0]));
+        }
+        catch (Exception e)
+        {
+            return Future.failedFuture("Invalid certificate passed");
```

do not swallow the exception.